### PR TITLE
[SPARK-23724][SPARK-23765][SQL] Line separator for the json datasource

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

Currently, [TextInputJsonDataSource](https://github.com/databricks/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonDataSource.scala#L86) uses [HadoopFileLinesReader](https://github.com/databricks/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonDataSource.scala#L125) to split json file to separate lines. The former one [splits](https://github.com/databricks/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonDataSource.scala#L125) json lines by [LineRecordReader](https://github.com/apache/hadoop/blob/trunk/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapreduce/lib/input/LineRecordReader.java#L68) without providing recordDelimiter. As a consequence of that, the hadoop library [reads lines terminated by one of CR, LF, or CRLF](https://github.com/apache/hadoop/blob/trunk/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/LineReader.java#L185-L254). The changes allow to specify the line separator instead of using the auto detection method of hadoop library.  If the separator is not specified, the line separation method of Hadoop is used by default.

## How was this patch tested?

Added new tests for writing/reading json files with custom line separator

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
@cloud-fan and @hvanhovell. Do you think we need the flexible option for line separator?

https://github.com/apache/spark/blob/bbff40206e6871ea9ab035e7a8876f495bdf3d90/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala#L91-L98

File: python/pyspark/sql/readwriter.py [hunk 1]
```diff
@@ -770,12 +773,15 @@ def json(self, path, mode=None, compression=None, dateFormat=None, timestampForm
                                 formats follow the formats at ``java.text.SimpleDateFormat``.
                                 This applies to timestamp type. If None is set, it uses the
                                 default value, ``yyyy-MM-dd'T'HH:mm:ss.SSSXXX``.
+        :param lineSep: defines the line separator that should be used for writing. If None is
+                        set, it uses the default value, ``\\n``.
```
Comment (cloud-fan) [inline python/pyspark/sql/readwriter.py]:
```
it covers all ``\\r``, ``\\r\\n`` and ``\\n``.
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala [hunk 2]
```diff
@@ -85,6 +85,38 @@ private[sql] class JSONOptions(
 
   val multiLine = parameters.get("multiLine").map(_.toBoolean).getOrElse(false)
 
+  val charset: Option[String] = Some("UTF-8")
+
+  /**
+   * A sequence of bytes between two consecutive json records. Format of the option is:
+   *   selector (1 char) + delimiter body (any length) | sequence of chars
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala]: I'm afraid of defining our own rule here, is there any standard we can follow?

File: python/pyspark/sql/readwriter.py (hunk 1 above)
Comment (MaxGekk) [reply python/pyspark/sql/readwriter.py]: It is a method of DataFrameWriter. It writes exactly `'\n'`

File: python/pyspark/sql/readwriter.py [hunk 3]
```diff
@@ -176,7 +176,7 @@ def json(self, path, schema=None, primitivesAsString=None, prefersDecimal=None,
              allowComments=None, allowUnquotedFieldNames=None, allowSingleQuotes=None,
              allowNumericLeadingZero=None, allowBackslashEscapingAnyCharacter=None,
              mode=None, columnNameOfCorruptRecord=None, dateFormat=None, timestampFormat=None,
-             multiLine=None, allowUnquotedControlChars=None):
+             multiLine=None, allowUnquotedControlChars=None, lineSep=None):
```
Comment (gatorsmile) [inline python/pyspark/sql/readwriter.py]: rename it to `recordDelimiter `

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala [hunk 4]
```diff
@@ -85,6 +85,38 @@ private[sql] class JSONOptions(
 
   val multiLine = parameters.get("multiLine").map(_.toBoolean).getOrElse(false)
 
+  val charset: Option[String] = Some("UTF-8")
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala]: It sounds like we need to review https://github.com/apache/spark/pull/20849 first

**Comment by MaxGekk** [comment]
Please, look at https://github.com/apache/spark/pull/20937

**Pull request closed without merging**